Add a `kubernetes-log4j` module

ppkarwasz opened the pull request:
## Description

The `kubernetes-log4j` module adds the ability to Log4j Core to use Kubernetes attributes in a configuration file.

It is a cleaned-up version of the [`org.apache.logging.log4j:log4j-kubernetes`](https://logging.apache.org/log4j/2.x/log4j-kubernetes.html), which is also released under ASL 2.0.

As explained in #5682, it does make more sense to host is here since:

 * it only depends on a very stable `StrLookup` dependency from `log4j-core`,
 * the number and kind of properties available through `kubernetes-client` depend on its version.

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

ppkarwasz commented:
For compatibility with `org.apache.logging.log4j:log4j-kubernetes` the lookup allows to override the automatic Kubernetes Client configuration using Log4j property sources. If this does not make sense in this project, I can remove the `ClientBuilder` and `ClientProperties` classes.

ppkarwasz commented:
I am converting this to draft, since I need to add tests for most of the code.

Do you have an util to replace the `ContainerUtil` from the original `log4j-kubernetes`? The class tries to determine the id of the Kubernetes container, by parsing `/proc/self/cgroup` on Linux, which is error prone and not easy to test.

manusa commented on log4j/pom.xml:
```diff
@@ -0,0 +1,89 @@
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <!-- Replace `jar:jar` execution with `bundle:bundle` -->
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>default-jar</id>
+            <phase>none</phase>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.felix</groupId>
+        <artifactId>maven-bundle-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>generate-osgi-bundle</id>
+            <goals>
+              <goal>bundle</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
```
I'm not 100% sure if these should be required as they are defined in the parent and you're already defining some of the osgi properties.

ppkarwasz replied:
I replaced the artifact type from `jar` to `bundle`, in order to activate the `maven-bundle-plugin`.

manusa commented:
>  If this does not make sense in this project, I can remove the ClientBuilder and ClientProperties classes.

I'm not sure how does this play with our own property inference logic:

https://github.com/fabric8io/kubernetes-client/blob/315f4869a0cfa658099270dfea0f4d143cc99899/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java#L409

I assume that for backwards-compatibility this is necessary, so I'm fine keeping it.

manusa commented:
> Do you have an util to replace the ContainerUtil from the original log4j-kubernetes?

Not that I know of.

manusa commented on log4j/src/test/java/io/fabric8/kubernetes/log4j/lookup/KubernetesLookupTest.java:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  private Namespace createNamespace() {
+    final Namespace namespace = new Namespace();
+    final ObjectMeta meta = new ObjectMeta();
+    final Map<String, String> annotations = new HashMap<>();
+    annotations.put("test", "name");
+    meta.setAnnotations(annotations);
+    final Map<String, String> labels = new HashMap<>();
+    labels.put("ns", "my-namespace");
+    meta.setLabels(labels);
+    meta.setUid(UUID.randomUUID().toString());
+    namespace.setMetadata(meta);
+    return namespace;
```
nit: You could use NamespaceBuilder fluent interface to make this a little bit more readable and avoid declaring intermediate variables (meta, annotations, and so on).

ppkarwasz replied:
Nice trick, thanks

manusa commented on pom.xml:
```diff
@@ -221,6 +222,7 @@
     <module>java-generator</module>
     <module>httpclient-okhttp</module>
     <module>httpclient-vertx</module>
+    <module>log4j</module>
```
To be able to properly generate our BOM this module should also be added as a dependency to the dependencyManagement section.

ppkarwasz replied:
Done

ppkarwasz commented on log4j/pom.xml:
```diff
@@ -0,0 +1,89 @@
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-core</artifactId>
+    </dependency>
```
I removed it.

ppkarwasz commented on pom.xml:
```diff
@@ -700,6 +702,15 @@
+        <artifactId>log4j-bom</artifactId>
+        <version>${log4j.version}</version>
+        <scope>import</scope>
+        <type>pom</type>
```
No, it is not necessary. I moved the only required dependency (`log4j-core`) to the module itself.

ppkarwasz commented:
@manusa,

Thank you for your review. I finally had some time to work on this PR.

Apart from addressing your remarks above:

- I introduced a system property `kubernetes.log4j.useProperties`. When set to `false` (default) Log4j properties are not used to configure the Kubernetes client. I think that this is a good compromise to maintain compatibility with `log4j-kubernetes`, but require users to set a special property to enable it.
- I increased the test coverage.

ppkarwasz commented:
The OpenShift failures seem unrelated to my PR.

rohanKanojia commented:
@ppkarwasz : Is it possible to add some README or user documentation in doc folder regarding how to use this?

ppkarwasz commented:
@rohanKanojia,

I have added a `doc/KubernetesLog4j.md` file in the `/doc` folder.

manusa reviewed (approved):
LGTM, thx!

manusa commented:
> @manusa,
> 
> Thank you for offering to maintain this component. If some Log4j-specific issues were to appear with it, feel free to `@`-mention me.

Will do :) Thanks for the contribution!
